<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single cell RNA-seq data</title>
  <meta name="description" content="Analysis of single cell RNA-seq data">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="Analysis of single cell RNA-seq data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis of single cell RNA-seq data" />
  
  
  

<meta name="author" content="Vladimir Kiselev (wikiselev), Tallulah Andrews, Davis McCarthy (davisjmcc), Maren BÃ¼ttner (marenbuettner) and Martin Hemberg (m_hemberg)">


<meta name="date" content="2017-10-27">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  

<link rel="next" href="introduction-to-single-cell-rna-seq.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-71525309-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#video"><i class="fa fa-check"></i><b>1.1</b> Video</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.3</b> GitHub</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#r-based"><i class="fa fa-check"></i><b>1.4</b> R-based</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#docker-image-rstudio"><i class="fa fa-check"></i><b>1.5</b> Docker image (RStudio)</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#manual-installation"><i class="fa fa-check"></i><b>1.6</b> Manual installation</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.7</b> License</a></li>
<li class="chapter" data-level="1.8" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.8</b> Prerequisites</a></li>
<li class="chapter" data-level="1.9" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.9</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html"><i class="fa fa-check"></i><b>2</b> Introduction to single-cell RNA-seq</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#bulk-rna-seq"><i class="fa fa-check"></i><b>2.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#scrna-seq"><i class="fa fa-check"></i><b>2.2</b> scRNA-seq</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#workflow"><i class="fa fa-check"></i><b>2.3</b> Workflow</a></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#computational-analysis"><i class="fa fa-check"></i><b>2.4</b> Computational Analysis</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#challenges"><i class="fa fa-check"></i><b>2.5</b> Challenges</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#experimental-methods"><i class="fa fa-check"></i><b>2.6</b> Experimental methods</a><ul>
<li class="chapter" data-level="2.6.1" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#overview-of-experimental-methods-for-generating-scrna-seq-data"><i class="fa fa-check"></i><b>2.6.1</b> Overview of experimental methods for generating scRNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-single-cell-rna-seq.html"><a href="introduction-to-single-cell-rna-seq.html#what-platform-to-use-for-my-experiment"><i class="fa fa-check"></i><b>2.7</b> What platform to use for my experiment?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html"><i class="fa fa-check"></i><b>3</b> Construction of expression matrix</a><ul>
<li class="chapter" data-level="3.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-qc"><i class="fa fa-check"></i><b>3.1</b> Reads QC</a></li>
<li class="chapter" data-level="3.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-alignment"><i class="fa fa-check"></i><b>3.2</b> Reads alignment</a></li>
<li class="chapter" data-level="3.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#alignment-example"><i class="fa fa-check"></i><b>3.3</b> Alignment example</a></li>
<li class="chapter" data-level="3.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-qc"><i class="fa fa-check"></i><b>3.4</b> Mapping QC</a></li>
<li class="chapter" data-level="3.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#reads-quantification"><i class="fa fa-check"></i><b>3.5</b> Reads quantification</a></li>
<li class="chapter" data-level="3.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#umichapter"><i class="fa fa-check"></i><b>3.6</b> Unique Molecular Identifiers (UMIs)</a><ul>
<li class="chapter" data-level="3.6.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction"><i class="fa fa-check"></i><b>3.6.1</b> Introduction</a></li>
<li class="chapter" data-level="3.6.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mapping-barcodes"><i class="fa fa-check"></i><b>3.6.2</b> Mapping Barcodes</a></li>
<li class="chapter" data-level="3.6.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#counting-barcodes"><i class="fa fa-check"></i><b>3.6.3</b> Counting Barcodes</a></li>
<li class="chapter" data-level="3.6.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correcting-for-errors"><i class="fa fa-check"></i><b>3.6.4</b> Correcting for Errors</a></li>
<li class="chapter" data-level="3.6.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downstream-analysis"><i class="fa fa-check"></i><b>3.6.5</b> Downstream Analysis</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor-singlecellexperiment-and-scater"><i class="fa fa-check"></i><b>3.7</b> Bioconductor, <code>SingleCellExperiment</code> and <code>scater</code></a><ul>
<li class="chapter" data-level="3.7.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#bioconductor"><i class="fa fa-check"></i><b>3.7.1</b> Bioconductor</a></li>
<li class="chapter" data-level="3.7.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#singlecellexperiment-class"><i class="fa fa-check"></i><b>3.7.2</b> <code>SingleCellExperiment</code> class</a></li>
<li class="chapter" data-level="3.7.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scater-package"><i class="fa fa-check"></i><b>3.7.3</b> <code>scater</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exprs-qc"><i class="fa fa-check"></i><b>3.8</b> Expression QC (UMI)</a><ul>
<li class="chapter" data-level="3.8.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-1"><i class="fa fa-check"></i><b>3.8.1</b> Introduction</a></li>
<li class="chapter" data-level="3.8.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tung-dataset"><i class="fa fa-check"></i><b>3.8.2</b> Tung dataset</a></li>
<li class="chapter" data-level="3.8.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-qc"><i class="fa fa-check"></i><b>3.8.3</b> Cell QC</a></li>
<li class="chapter" data-level="3.8.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cell-filtering"><i class="fa fa-check"></i><b>3.8.4</b> Cell filtering</a></li>
<li class="chapter" data-level="3.8.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#compare-filterings"><i class="fa fa-check"></i><b>3.8.5</b> Compare filterings</a></li>
<li class="chapter" data-level="3.8.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#gene-analysis"><i class="fa fa-check"></i><b>3.8.6</b> Gene analysis</a></li>
<li class="chapter" data-level="3.8.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#save-the-data"><i class="fa fa-check"></i><b>3.8.7</b> Save the data</a></li>
<li class="chapter" data-level="3.8.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise"><i class="fa fa-check"></i><b>3.8.8</b> Big Exercise</a></li>
<li class="chapter" data-level="3.8.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo"><i class="fa fa-check"></i><b>3.8.9</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#expression-qc-reads"><i class="fa fa-check"></i><b>3.9</b> Expression QC (Reads)</a><ul>
<li class="chapter" data-level="3.9.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-1"><i class="fa fa-check"></i><b>3.9.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization"><i class="fa fa-check"></i><b>3.10</b> Data visualization</a><ul>
<li class="chapter" data-level="3.10.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-2"><i class="fa fa-check"></i><b>3.10.1</b> Introduction</a></li>
<li class="chapter" data-level="3.10.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-pca"><i class="fa fa-check"></i><b>3.10.2</b> PCA plot</a></li>
<li class="chapter" data-level="3.10.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#visual-tsne"><i class="fa fa-check"></i><b>3.10.3</b> tSNE map</a></li>
<li class="chapter" data-level="3.10.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#big-exercise-1"><i class="fa fa-check"></i><b>3.10.4</b> Big Exercise</a></li>
<li class="chapter" data-level="3.10.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-2"><i class="fa fa-check"></i><b>3.10.5</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#data-visualization-reads"><i class="fa fa-check"></i><b>3.11</b> Data visualization (Reads)</a><ul>
<li class="chapter" data-level="3.11.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-3"><i class="fa fa-check"></i><b>3.11.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.12" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors"><i class="fa fa-check"></i><b>3.12</b> Identifying confounding factors</a><ul>
<li class="chapter" data-level="3.12.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-3"><i class="fa fa-check"></i><b>3.12.1</b> Introduction</a></li>
<li class="chapter" data-level="3.12.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#correlations-with-pcs"><i class="fa fa-check"></i><b>3.12.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="3.12.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#explanatory-variables"><i class="fa fa-check"></i><b>3.12.3</b> Explanatory variables</a></li>
<li class="chapter" data-level="3.12.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#other-confounders"><i class="fa fa-check"></i><b>3.12.4</b> Other confounders</a></li>
<li class="chapter" data-level="3.12.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise"><i class="fa fa-check"></i><b>3.12.5</b> Exercise</a></li>
<li class="chapter" data-level="3.12.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-4"><i class="fa fa-check"></i><b>3.12.6</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.13" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#identifying-confounding-factors-reads"><i class="fa fa-check"></i><b>3.13</b> Identifying confounding factors (Reads)</a><ul>
<li class="chapter" data-level="3.13.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-5"><i class="fa fa-check"></i><b>3.13.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.14" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-theory"><i class="fa fa-check"></i><b>3.14</b> Normalization theory</a><ul>
<li class="chapter" data-level="3.14.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-4"><i class="fa fa-check"></i><b>3.14.1</b> Introduction</a></li>
<li class="chapter" data-level="3.14.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#library-size-1"><i class="fa fa-check"></i><b>3.14.2</b> Library size</a></li>
<li class="chapter" data-level="3.14.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalisations"><i class="fa fa-check"></i><b>3.14.3</b> Normalisations</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-umi"><i class="fa fa-check"></i><b>3.15</b> Normalization practice (UMI)</a><ul>
<li class="chapter" data-level="3.15.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#raw"><i class="fa fa-check"></i><b>3.15.1</b> Raw</a></li>
<li class="chapter" data-level="3.15.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#cpm"><i class="fa fa-check"></i><b>3.15.2</b> CPM</a></li>
<li class="chapter" data-level="3.15.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#tmm"><i class="fa fa-check"></i><b>3.15.3</b> TMM</a></li>
<li class="chapter" data-level="3.15.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#scran"><i class="fa fa-check"></i><b>3.15.4</b> scran</a></li>
<li class="chapter" data-level="3.15.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#size-factor-rle"><i class="fa fa-check"></i><b>3.15.5</b> Size-factor (RLE)</a></li>
<li class="chapter" data-level="3.15.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#upperquantile"><i class="fa fa-check"></i><b>3.15.6</b> Upperquantile</a></li>
<li class="chapter" data-level="3.15.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#downsampling"><i class="fa fa-check"></i><b>3.15.7</b> Downsampling</a></li>
<li class="chapter" data-level="3.15.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalizing-for-genetranscript-length"><i class="fa fa-check"></i><b>3.15.8</b> Normalizing for gene/transcript length</a></li>
<li class="chapter" data-level="3.15.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise-1"><i class="fa fa-check"></i><b>3.15.9</b> Exercise</a></li>
<li class="chapter" data-level="3.15.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-6"><i class="fa fa-check"></i><b>3.15.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.16" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#normalization-practice-reads"><i class="fa fa-check"></i><b>3.16</b> Normalization practice (Reads)</a><ul>
<li class="chapter" data-level="3.16.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-7"><i class="fa fa-check"></i><b>3.16.1</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.17" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#dealing-with-confounders"><i class="fa fa-check"></i><b>3.17</b> Dealing with confounders</a><ul>
<li class="chapter" data-level="3.17.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#introduction-5"><i class="fa fa-check"></i><b>3.17.1</b> Introduction</a></li>
<li class="chapter" data-level="3.17.2" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#remove-unwanted-variation"><i class="fa fa-check"></i><b>3.17.2</b> Remove Unwanted Variation</a></li>
<li class="chapter" data-level="3.17.3" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#combat"><i class="fa fa-check"></i><b>3.17.3</b> Combat</a></li>
<li class="chapter" data-level="3.17.4" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#mnncorrect"><i class="fa fa-check"></i><b>3.17.4</b> mnnCorrect</a></li>
<li class="chapter" data-level="3.17.5" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#glm"><i class="fa fa-check"></i><b>3.17.5</b> GLM</a></li>
<li class="chapter" data-level="3.17.6" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-1"><i class="fa fa-check"></i><b>3.17.6</b> Effectiveness 1</a></li>
<li class="chapter" data-level="3.17.7" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-2"><i class="fa fa-check"></i><b>3.17.7</b> Effectiveness 2</a></li>
<li class="chapter" data-level="3.17.8" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-3"><i class="fa fa-check"></i><b>3.17.8</b> Effectiveness 3</a></li>
<li class="chapter" data-level="3.17.9" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#effectiveness-4"><i class="fa fa-check"></i><b>3.17.9</b> Effectiveness 4</a></li>
<li class="chapter" data-level="3.17.10" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#exercise-2"><i class="fa fa-check"></i><b>3.17.10</b> Exercise</a></li>
<li class="chapter" data-level="3.17.11" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#sessioninfo-8"><i class="fa fa-check"></i><b>3.17.11</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="3.18" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#dealing-with-confounders-reads"><i class="fa fa-check"></i><b>3.18</b> Dealing with confounders (Reads)</a><ul>
<li class="chapter" data-level="3.18.1" data-path="construction-of-expression-matrix.html"><a href="construction-of-expression-matrix.html#remove-unwanted-variation-1"><i class="fa fa-check"></i><b>3.18.1</b> Remove Unwanted Variation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="biological-analysis.html"><a href="biological-analysis.html"><i class="fa fa-check"></i><b>4</b> Biological Analysis</a><ul>
<li class="chapter" data-level="4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-introduction"><i class="fa fa-check"></i><b>4.1</b> Clustering Introduction</a><ul>
<li class="chapter" data-level="4.1.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-6"><i class="fa fa-check"></i><b>4.1.1</b> Introduction</a></li>
<li class="chapter" data-level="4.1.2" data-path="biological-analysis.html"><a href="biological-analysis.html#dimensionality-reductions"><i class="fa fa-check"></i><b>4.1.2</b> Dimensionality reductions</a></li>
<li class="chapter" data-level="4.1.3" data-path="biological-analysis.html"><a href="biological-analysis.html#clustering-methods"><i class="fa fa-check"></i><b>4.1.3</b> Clustering methods</a></li>
<li class="chapter" data-level="4.1.4" data-path="biological-analysis.html"><a href="biological-analysis.html#challenges-in-clustering"><i class="fa fa-check"></i><b>4.1.4</b> Challenges in clustering</a></li>
<li class="chapter" data-level="4.1.5" data-path="biological-analysis.html"><a href="biological-analysis.html#tools-for-scrna-seq-data"><i class="fa fa-check"></i><b>4.1.5</b> Tools for scRNA-seq data</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#clust-methods"><i class="fa fa-check"></i><b>4.2</b> Clustering example</a><ul>
<li class="chapter" data-level="4.2.1" data-path="biological-analysis.html"><a href="biological-analysis.html#deng-dataset"><i class="fa fa-check"></i><b>4.2.1</b> Deng dataset</a></li>
<li class="chapter" data-level="4.2.2" data-path="biological-analysis.html"><a href="biological-analysis.html#sc3-1"><i class="fa fa-check"></i><b>4.2.2</b> SC3</a></li>
<li class="chapter" data-level="4.2.3" data-path="biological-analysis.html"><a href="biological-analysis.html#pcareduce-1"><i class="fa fa-check"></i><b>4.2.3</b> pcaReduce</a></li>
<li class="chapter" data-level="4.2.4" data-path="biological-analysis.html"><a href="biological-analysis.html#tsne-kmeans"><i class="fa fa-check"></i><b>4.2.4</b> tSNE + kmeans</a></li>
<li class="chapter" data-level="4.2.5" data-path="biological-analysis.html"><a href="biological-analysis.html#snn-cliq-1"><i class="fa fa-check"></i><b>4.2.5</b> SNN-Cliq</a></li>
<li class="chapter" data-level="4.2.6" data-path="biological-analysis.html"><a href="biological-analysis.html#sincera-1"><i class="fa fa-check"></i><b>4.2.6</b> SINCERA</a></li>
<li class="chapter" data-level="4.2.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-10"><i class="fa fa-check"></i><b>4.2.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#feature-selection"><i class="fa fa-check"></i><b>4.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="4.3.1" data-path="biological-analysis.html"><a href="biological-analysis.html#identifying-genes-vs-a-null-model"><i class="fa fa-check"></i><b>4.3.1</b> Identifying Genes vs a Null Model</a></li>
<li class="chapter" data-level="4.3.2" data-path="biological-analysis.html"><a href="biological-analysis.html#correlated-expression"><i class="fa fa-check"></i><b>4.3.2</b> Correlated Expression</a></li>
<li class="chapter" data-level="4.3.3" data-path="biological-analysis.html"><a href="biological-analysis.html#comparing-methods"><i class="fa fa-check"></i><b>4.3.3</b> Comparing Methods</a></li>
<li class="chapter" data-level="4.3.4" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-11"><i class="fa fa-check"></i><b>4.3.4</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#pseudotime-analysis"><i class="fa fa-check"></i><b>4.4</b> Pseudotime analysis</a><ul>
<li class="chapter" data-level="4.4.1" data-path="biological-analysis.html"><a href="biological-analysis.html#tscan"><i class="fa fa-check"></i><b>4.4.1</b> TSCAN</a></li>
<li class="chapter" data-level="4.4.2" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle"><i class="fa fa-check"></i><b>4.4.2</b> monocle</a></li>
<li class="chapter" data-level="4.4.3" data-path="biological-analysis.html"><a href="biological-analysis.html#diffusion-maps"><i class="fa fa-check"></i><b>4.4.3</b> Diffusion maps</a></li>
<li class="chapter" data-level="4.4.4" data-path="biological-analysis.html"><a href="biological-analysis.html#slicer"><i class="fa fa-check"></i><b>4.4.4</b> SLICER</a></li>
<li class="chapter" data-level="4.4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>4.4.5</b> Comparison of the methods</a></li>
<li class="chapter" data-level="4.4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#expression-of-genes-through-time"><i class="fa fa-check"></i><b>4.4.6</b> Expression of genes through time</a></li>
<li class="chapter" data-level="4.4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-12"><i class="fa fa-check"></i><b>4.4.7</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="biological-analysis.html"><a href="biological-analysis.html#imputation"><i class="fa fa-check"></i><b>4.5</b> Imputation</a><ul>
<li class="chapter" data-level="4.5.1" data-path="biological-analysis.html"><a href="biological-analysis.html#scimpute"><i class="fa fa-check"></i><b>4.5.1</b> scImpute</a></li>
<li class="chapter" data-level="4.5.2" data-path="biological-analysis.html"><a href="biological-analysis.html#magic"><i class="fa fa-check"></i><b>4.5.2</b> MAGIC</a></li>
<li class="chapter" data-level="4.5.3" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-13"><i class="fa fa-check"></i><b>4.5.3</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="biological-analysis.html"><a href="biological-analysis.html#dechapter"><i class="fa fa-check"></i><b>4.6</b> Differential Expression (DE) analysis</a><ul>
<li class="chapter" data-level="4.6.1" data-path="biological-analysis.html"><a href="biological-analysis.html#bulk-rna-seq-1"><i class="fa fa-check"></i><b>4.6.1</b> Bulk RNA-seq</a></li>
<li class="chapter" data-level="4.6.2" data-path="biological-analysis.html"><a href="biological-analysis.html#single-cell-rna-seq"><i class="fa fa-check"></i><b>4.6.2</b> Single cell RNA-seq</a></li>
<li class="chapter" data-level="4.6.3" data-path="biological-analysis.html"><a href="biological-analysis.html#differences-in-distribution"><i class="fa fa-check"></i><b>4.6.3</b> Differences in Distribution</a></li>
<li class="chapter" data-level="4.6.4" data-path="biological-analysis.html"><a href="biological-analysis.html#models-of-single-cell-rnaseq-data"><i class="fa fa-check"></i><b>4.6.4</b> Models of single-cell RNASeq data</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="biological-analysis.html"><a href="biological-analysis.html#de-in-a-real-dataset"><i class="fa fa-check"></i><b>4.7</b> DE in a real dataset</a><ul>
<li class="chapter" data-level="4.7.1" data-path="biological-analysis.html"><a href="biological-analysis.html#introduction-7"><i class="fa fa-check"></i><b>4.7.1</b> Introduction</a></li>
<li class="chapter" data-level="4.7.2" data-path="biological-analysis.html"><a href="biological-analysis.html#kolmogorov-smirnov-test"><i class="fa fa-check"></i><b>4.7.2</b> Kolmogorov-Smirnov test</a></li>
<li class="chapter" data-level="4.7.3" data-path="biological-analysis.html"><a href="biological-analysis.html#wilcoxmann-whitney-u-test"><i class="fa fa-check"></i><b>4.7.3</b> Wilcox/Mann-Whitney-U Test</a></li>
<li class="chapter" data-level="4.7.4" data-path="biological-analysis.html"><a href="biological-analysis.html#edger"><i class="fa fa-check"></i><b>4.7.4</b> edgeR</a></li>
<li class="chapter" data-level="4.7.5" data-path="biological-analysis.html"><a href="biological-analysis.html#monocle-1"><i class="fa fa-check"></i><b>4.7.5</b> Monocle</a></li>
<li class="chapter" data-level="4.7.6" data-path="biological-analysis.html"><a href="biological-analysis.html#mast"><i class="fa fa-check"></i><b>4.7.6</b> MAST</a></li>
<li class="chapter" data-level="4.7.7" data-path="biological-analysis.html"><a href="biological-analysis.html#slow-methods-1h-to-run"><i class="fa fa-check"></i><b>4.7.7</b> Slow Methods (&gt;1h to run)</a></li>
<li class="chapter" data-level="4.7.8" data-path="biological-analysis.html"><a href="biological-analysis.html#bpsc"><i class="fa fa-check"></i><b>4.7.8</b> BPSC</a></li>
<li class="chapter" data-level="4.7.9" data-path="biological-analysis.html"><a href="biological-analysis.html#scde"><i class="fa fa-check"></i><b>4.7.9</b> SCDE</a></li>
<li class="chapter" data-level="4.7.10" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-14"><i class="fa fa-check"></i><b>4.7.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="biological-analysis.html"><a href="biological-analysis.html#projecting-scrna-seq-data"><i class="fa fa-check"></i><b>4.8</b> Projecting scRNA-seq data</a><ul>
<li class="chapter" data-level="4.8.1" data-path="biological-analysis.html"><a href="biological-analysis.html#datasets"><i class="fa fa-check"></i><b>4.8.1</b> Datasets</a></li>
<li class="chapter" data-level="4.8.2" data-path="biological-analysis.html"><a href="biological-analysis.html#run-scmap"><i class="fa fa-check"></i><b>4.8.2</b> Run <code>scmap</code></a></li>
<li class="chapter" data-level="4.8.3" data-path="biological-analysis.html"><a href="biological-analysis.html#results"><i class="fa fa-check"></i><b>4.8.3</b> Results</a></li>
<li class="chapter" data-level="4.8.4" data-path="biological-analysis.html"><a href="biological-analysis.html#creating-a-precomputed-reference"><i class="fa fa-check"></i><b>4.8.4</b> Creating a precomputed Reference</a></li>
<li class="chapter" data-level="4.8.5" data-path="biological-analysis.html"><a href="biological-analysis.html#sessioninfo-15"><i class="fa fa-check"></i><b>4.8.5</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="biological-analysis.html"><a href="biological-analysis.html#seurat"><i class="fa fa-check"></i><b>5</b> <code>Seurat</code></a><ul>
<li class="chapter" data-level="5.1" data-path="seurat.html"><a href="seurat.html"><i class="fa fa-check"></i><b>5.1</b> <code>seurat</code> object class</a></li>
<li class="chapter" data-level="5.2" data-path="seurat.html"><a href="seurat.html#expression-qc"><i class="fa fa-check"></i><b>5.2</b> Expression QC</a></li>
<li class="chapter" data-level="5.3" data-path="seurat.html"><a href="seurat.html#normalization"><i class="fa fa-check"></i><b>5.3</b> Normalization</a></li>
<li class="chapter" data-level="5.4" data-path="seurat.html"><a href="seurat.html#highly-variable-genes-1"><i class="fa fa-check"></i><b>5.4</b> Highly variable genes</a></li>
<li class="chapter" data-level="5.5" data-path="seurat.html"><a href="seurat.html#dealing-with-confounders-1"><i class="fa fa-check"></i><b>5.5</b> Dealing with confounders</a></li>
<li class="chapter" data-level="5.6" data-path="seurat.html"><a href="seurat.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>5.6</b> Linear dimensionality reduction</a></li>
<li class="chapter" data-level="5.7" data-path="seurat.html"><a href="seurat.html#significant-pcs"><i class="fa fa-check"></i><b>5.7</b> Significant PCs</a></li>
<li class="chapter" data-level="5.8" data-path="seurat.html"><a href="seurat.html#clustering-cells"><i class="fa fa-check"></i><b>5.8</b> Clustering cells</a></li>
<li class="chapter" data-level="5.9" data-path="seurat.html"><a href="seurat.html#marker-genes"><i class="fa fa-check"></i><b>5.9</b> Marker genes</a></li>
<li class="chapter" data-level="5.10" data-path="seurat.html"><a href="seurat.html#sessioninfo-16"><i class="fa fa-check"></i><b>5.10</b> sessionInfo()</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><i class="fa fa-check"></i><b>6</b> âIdealâ scRNAseq pipeline (as of Mar 2017)</a><ul>
<li class="chapter" data-level="6.1" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#experimental-design"><i class="fa fa-check"></i><b>6.1</b> Experimental Design</a></li>
<li class="chapter" data-level="6.2" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#processing-reads"><i class="fa fa-check"></i><b>6.2</b> Processing Reads</a></li>
<li class="chapter" data-level="6.3" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#preparing-expression-matrix"><i class="fa fa-check"></i><b>6.3</b> Preparing Expression Matrix</a></li>
<li class="chapter" data-level="6.4" data-path="ideal-scrnaseq-pipeline-as-of-mar-2017.html"><a href="ideal-scrnaseq-pipeline-as-of-mar-2017.html#biological-interpretation"><i class="fa fa-check"></i><b>6.4</b> Biological Interpretation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-exercises.html"><a href="advanced-exercises.html"><i class="fa fa-check"></i><b>7</b> Advanced exercises</a></li>
<li class="chapter" data-level="8" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>8</b> Resources</a><ul>
<li class="chapter" data-level="8.1" data-path="resources.html"><a href="resources.html#scrna-seq-protocols"><i class="fa fa-check"></i><b>8.1</b> scRNA-seq protocols</a></li>
<li class="chapter" data-level="8.2" data-path="resources.html"><a href="resources.html#external-rna-control-consortium-ercc"><i class="fa fa-check"></i><b>8.2</b> External RNA Control Consortium (ERCC)</a></li>
<li class="chapter" data-level="8.3" data-path="resources.html"><a href="resources.html#scrna-seq-analysis-tools"><i class="fa fa-check"></i><b>8.3</b> scRNA-seq analysis tools</a></li>
<li class="chapter" data-level="8.4" data-path="resources.html"><a href="resources.html#scrna-seq-public-datasets"><i class="fa fa-check"></i><b>8.4</b> scRNA-seq public datasets</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>9</b> References</a></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab, 2017</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single cell RNA-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Analysis of single cell RNA-seq data</h1>
<h4 class="author"><em>Vladimir Kiselev (<a href = 'https://twitter.com/wikiselev'>wikiselev</a>), Tallulah Andrews, Davis McCarthy (<a href = 'https://twitter.com/davisjmcc'>davisjmcc</a>), Maren BÃ¼ttner (<a href = 'https://twitter.com/marenbuettner'>marenbuettner</a>) and Martin Hemberg (<a href = 'https://twitter.com/m_hemberg'>m_hemberg</a>)</em></h4>
<h4 class="date"><em>2017-10-27</em></h4>
</div>
<div id="about-the-course" class="section level1">
<h1><span class="header-section-number">1</span> About the course</h1>
<blockquote>
<p><span style="color:red"><strong>Important!</strong> The course will be run on the <strong>31st October - 1st November 2017, both days 9:00-17:00 London, UK time</strong>. It will also be <strong>live broadcast</strong> at the <a href="https://www.youtube.com/channel/UCsc6r6UKxb2qRcDQPix2L5A">Bioinformatics Training channel</a> on YouTube. Please join the broadcast!</span></p>
</blockquote>
<p>Today it is possible to obtain genome-wide transcriptome data from single cells using high-throughput sequencing (scRNA-seq). The main advantage of scRNA-seq is that the cellular resolution and the genome wide scope makes it possible to address issues that are intractable using other methods, e.g.Â bulk RNA-seq or single-cell RT-qPCR. However, to analyze scRNA-seq data, novel methods are required and some of the underlying assumptions for the methods developed for bulk RNA-seq experiments are no longer valid.</p>
<p>In this course we will discuss some of the questions that can be addressed using scRNA-seq as well as the available computational and statistical methods avialable. The course is taught through the University of Cambridge <a href="http://training.csx.cam.ac.uk/bioinformatics/" target="blank">Bioinformatics training unit</a>, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data. The course is taught twice per year and the material here is updated prior to each event.</p>
<p>The number of computational tools is increasing rapidly and we are doing our best to keep up to date with what is available. One of the main constraints for this course is that we would like to use tools that are implemented in R and that run reasonably fast. Moreover, we will also confess to being somewhat biased towards methods that have been developed either by us or by our friends and colleagues.</p>
<div id="video" class="section level2">
<h2><span class="header-section-number">1.1</span> Video</h2>
<iframe width="560" height="315" src="https://www.youtube.com/embed/i58Fk6R03PA?list=PLEyKDyF1qdOaoWBu8jNwN4o5z6wlm3aES" frameborder="0" allowfullscreen>
</iframe>
</div>
<div id="registration" class="section level2">
<h2><span class="header-section-number">1.2</span> Registration</h2>
<p>Please follow this link and register for the <strong>âAnalysis of single cell RNA-seq dataâ</strong> course: <a href="http://training.csx.cam.ac.uk/bioinformatics/search" target="blank">http://training.csx.cam.ac.uk/bioinformatics/search</a></p>
</div>
<div id="github" class="section level2">
<h2><span class="header-section-number">1.3</span> GitHub</h2>
<p><a href="https://github.com/hemberg-lab/scRNA.seq.course" target="blank">https://github.com/hemberg-lab/scRNA.seq.course</a></p>
</div>
<div id="r-based" class="section level2">
<h2><span class="header-section-number">1.4</span> R-based</h2>
<p><a href="https://www.r-project.org/">R</a> is one of the most popular programming languages for Bioinformatics. We aimed to only use R packages in this course. However, a few tools that we describe (<a href="http://bioinfo.uncc.edu/SNNCliq/">SNN-Cliq</a> and <a href="https://github.com/pkathail/magic">MAGIC</a>) are Python-based.</p>
</div>
<div id="docker-image-rstudio" class="section level2">
<h2><span class="header-section-number">1.5</span> Docker image (RStudio)</h2>
<p>The course can be reproduced without any package installation by running the course docker RStudio image which contains all the required packages.</p>
<p>Make sure Docker is installed on your system. If not, please follow <a href="https://docs.docker.com/engine/installation/">these instructions</a>. To run the course RStudio docker image:</p>
<pre><code>docker run -d -p 8787:8787 quay.io/hemberg-group/scrna-seq-course-rstudio</code></pre>
<p>This download the docker image (may take some time) and start a new Rstudio session in a docker container with all packages installed and all data files available.</p>
<p>Then visit <code>localhost:8787</code> in your browser and log in with <code>username:password</code> as <code>rstudio:rstudio</code>. Now you are ready to go!</p>
<p>More details on how ot run RStudio docker with different options can be found <a href="https://hub.docker.com/r/rocker/rstudio-stable/">here</a>.</p>
</div>
<div id="manual-installation" class="section level2">
<h2><span class="header-section-number">1.6</span> Manual installation</h2>
<p>If you are not using a docker image of the course, then to be able to run all code chunks of the course you need to clone or download the <a href="https://github.com/hemberg-lab/scRNA.seq.course">course GitHub repository</a> and start an R session in the cloned folder. You will also need to install all packages listed in the course <a href="https://github.com/hemberg-lab/scRNA.seq.course/blob/master/Dockerfile">Dockerfile</a>.</p>
<p>Alternatively, you can just install packages listed in a chapter of interest.</p>
</div>
<div id="license" class="section level2">
<h2><span class="header-section-number">1.7</span> License</h2>
<p>All of the course material is licensed under <b>GPL-3</b>. Anyone is welcome to go through the material in order to learn about analysis of scRNA-seq data. If you plan to use the material for your own teaching, we would appreciate if you tell us about it in addition to providing a suitable citation.</p>
</div>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">1.8</span> Prerequisites</h2>
<p>The course is intended for those who have basic familiarity with Unix and the R scripting language.</p>
<p>We will also assume that you are familiar with mapping and analysing bulk RNA-seq data as well as with the commonly available computational tools.</p>
<p>We recommend attending the <a href="http://training.csx.cam.ac.uk/bioinformatics/search">Introduction to RNA-seq and ChIP-seq data analysis</a> or the <a href="http://training.csx.cam.ac.uk/bioinformatics/search">Analysis of high-throughput sequencing data with Bioconductor</a> before attending this course.</p>
</div>
<div id="contact" class="section level2">
<h2><span class="header-section-number">1.9</span> Contact</h2>
<p>If you have any <strong>comments</strong>, <strong>questions</strong> or <strong>suggestions</strong> about the material, please contact <a href="mailto:vladimir.yu.kiselev@gmail.com">Vladimir Kiselev</a>.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="introduction-to-single-cell-rna-seq.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/hemberg-lab/scRNA.seq.course/edit/master/index.Rmd",
"text": "Edit"
},
"download": ["scRNA-seq-course.pdf"],
"toc": {
"collapse": "section"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
